### PR TITLE
Add workload config command

### PR DESCRIPTION
**Pull request by dsplaisted**
- Add `dotnet workload config` command to set the workload update mode
- Fix displaying the workload version in `dotnet --info` and `dotnet workload --version` when in workload set mode but no workload set is installed
- Add some initial workload set MSI installation tests

Here's the how the new `dotnet workload config` command works:

- `dotnet workload config --update-mode` - Prints the current workload update mode
- `dotnet workload config --update-mode workload-set` - Switches to workload set update mode
- `dotnet workload config --update-mode manifests` - Switches to loose manifest update mode

Here's the command-line help:

```
C:\Users\User>dotnet workload config --help
Description:
  Modify or display workload configuration values.
  To display a value, specify the corresponding command-line option without providing a value.  For example: "dotnet
  workload config --update-mode"

Usage:
  dotnet workload config [options]

Options:
  --update-mode <manifests|workload-set>  Controls whether updates should look for workload sets or the latest version
                                          of each individual manifest.
  -?, -h, --help                          Show command line help.
```

This replaces the previous syntax, which was `dotnet workload update --mode workloadset` and `dotnet workload update --mode loosemanifest`.  Note that the previous option was hidden from command-line help, where currently the `dotnet workload config` command is not hidden.

### CODE REVIEW

**Comment by baronfel** [comment]
For these kinds of config-lookup commands there are generally two patterns:

* [NPM-style](https://docs.npmjs.com/cli/v10/commands/npm-config), where you have a syntax like `npm config <get/set/delete/list> <key>[=optional value if the subcommand requires it]`
  * Rust's `cargo` also uses this format
* [Git style](https://www.atlassian.com/git/tutorials/setting-up-a-repository/git-config), where you have `git config <key> [value]`, where the presence of `value` signals a write instead of a read

I think using one of these forms would be easier for users to understand, because they would have experience in that form from other tools. This form would also scale to a more general `dotnet config` command in the future, where workload-specific config items would become something like `workload.update-mode` instead of just `update-mode`, for example.

To me, we should use one of these forms and use CliArguments to model these inputs. The CliArguments can use the same validation as currently described but make it easier to detect if the argument was even provided (can be controlled via Arity) and the values can be validated once the settings key is known using a validator in much the same way that you have already implemented.

In teams chat today, @marcpopMSFT mentioned that rollback and download link printing could be moved to this config command - I could see that, but I could also see versions like:
* a read-only settings key with a name like `package-links` or `rollback-file`
* a version of `dotnet workload --info` that accepted a flag to print these out, instead of update

I don't see as clear of an answer for that question.

**Comment by MiYanni** [comment]
@baronfel Correct me if I'm wrong, but doesn't the dotnet CLI have its own style of command structure? Like, a general command format that most of the commands use. I know we don't have rules/regulations against the command format. But I think your comment makes sense in the context of a brand new `dotnet config` command, where that command would have its own design identity.

In the context of just adding a small set of options for `dotnet workload`, I recommended we keep it feeling like the style of the rest of the `dotnet workload` commands. I think that'll make the most sense for this PR. And if we want to eventually make a `dotnet config` generalized command, we can add the ability to update the workload configuration there and eventually deprecate this usage in `dotnet workload`.

**Comment by baronfel** [comment]
We talked today and we're going to keep the existing format in this PR, keeping a more broad treatment of the CLI grammar of a config command for a larger, centralized config effort.

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            {
+                if (_updateMode == WorkloadConfigCommandParser.UpdateMode_WorkloadSet)
+                {
+                    _workloadInstaller.UpdateInstallMode(_sdkFeatureBand, true);
+                }
+                else if (_updateMode == WorkloadConfigCommandParser.UpdateMode_Manifests)
+                {
+                    _workloadInstaller.UpdateInstallMode(_sdkFeatureBand, false);
+                }
+                else if (string.IsNullOrEmpty(_updateMode))
```
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:70]: The _hasUpdateMode check doesn't eliminate this possibility?

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+        public override int Execute()
+        {
+            if (_hasUpdateMode)
+            {
+                if (_updateMode == WorkloadConfigCommandParser.UpdateMode_WorkloadSet)
+                {
+                    _workloadInstaller.UpdateInstallMode(_sdkFeatureBand, true);
+                }
+                else if (_updateMode == WorkloadConfigCommandParser.UpdateMode_Manifests)
```
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:66]: .Equals for strings, no?

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommandParser.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+using System.CommandLine;
+using Microsoft.DotNet.Workloads.Workload.Config;
+
+namespace Microsoft.DotNet.Cli
+{
+    internal static class WorkloadConfigCommandParser
+    {
+        //  dotnet workload config --update-mode workload-set
+
+        public static readonly string UpdateMode_WorkloadSet = "workload-set";
```
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommandParser.cs:13]: nit: typing - is hard. workloadset?

Also, I thought we were trying to not include "workload set" terminology? Not that I have a better proposal in this case...

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommandParser.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        public static readonly string UpdateMode_Manifests = "manifests";
+
+        public static readonly CliOption<string> UpdateMode = new("--update-mode")
+        {
+            Description = LocalizableStrings.UpdateModeDescription,
+            //Hidden = true,
```
Comment (Forgind) [inline src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommandParser.cs:19]: I think not hidden is reasonable. Delete?

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs [hunk 5]
```diff
@@ -159,11 +159,6 @@ public void RefreshWorkloadManifests()
                 return _workloadSet?.Version!;
             }
 
-            if (InstallStateContents.FromPath(Path.Combine(WorkloadInstallType.GetInstallStateFolder(_sdkVersionBand, _sdkRootPath), "default.json")).UseWorkloadSets == true)
```
Comment (Forgind) [inline src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:162]: We'd want this to change to echo the version for the current SDK, per our discussion this morning.

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.NET.Sdk.WorkloadManifestReader;
+
+#nullable enable
+
+namespace Microsoft.DotNet.Workloads.Workload.Config
+{
+    internal class WorkloadConfigCommand : WorkloadCommandBase
+    {
+        bool _hasUpdateMode;
+        string? _updateMode;
```
Comment (MiYanni) [inline src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:24]: nit `private`

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+        readonly IInstaller _workloadInstaller;
+
+        public WorkloadConfigCommand(
+            ParseResult parseResult,
+            IReporter? reporter = null,
+            IWorkloadResolverFactory? workloadResolverFactory = null
+        ) : base(parseResult, CommonOptions.HiddenVerbosityOption, reporter)
+        {
+            //  TODO: Is it possible to check the order of the options?  This would allow us to print the values out in the same order they are specified on the command line
```
Comment (MiYanni) [inline src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:41]: You can only get the list of tokens via the `Tokens` property. The list *should* be in the order in which they are read. But, that doesn't have a relationship with the `CliArguments` in any way, unfortunately. You'd have to make custom logic to figure that out.

**Review by MiYanni — APPROVED** [review]

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs (hunk 6 above)
Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:24]: `private` would be redundant here.  I know it's inconsistent here and probably is throughout our codebase.  Do we have a convention we are trying to follow for whether we should include `private` or not?

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs (hunk 2 above)
Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:66]: Fixed (also with invariant ignore case)

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs (hunk 1 above)
Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs:70]: No, `_hasUpdateMode` specifies whether the `--update-mode` option was specified, but it's possible to specify the option without it's optional argument value, which we use to print the current value.

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommandParser.cs (hunk 3 above)
Comment (dsplaisted) [reply src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommandParser.cs:13]: I'm pretty sure the guidance I got from @baronfel was to use dashes to separate words.  I think that's the style we're going with in the CLI.

File: src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs (hunk 5 above)
Comment (dsplaisted) [reply src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs:162]: I've filed an issue to track the further changes needed: https://github.com/dotnet/sdk/issues/40005

File: src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs (hunk 6 above)
Comment (Forgind) [reply src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs]: I generally prefer including it...not technically an answer, but it's my preference :)

Comment (MiYanni) [reply src/Cli/dotnet/commands/dotnet-workload/config/WorkloadConfigCommand.cs]: I'm pretty sure explicit access modifiers for fields are the convention in the dotnet org. I'm surprised we don't have that enabled in our editorconfig.

**Pull request merged**